File: custom-scripts/ori/ori.py
#!/usr/bin/env uv run --script
# /// script
# requires-python = ">=3.12"
# dependencies = [
#     "flask",
#     "requests",
# ]
# ///

import subprocess
import requests
import os
import json
from flask import Flask, request, jsonify

app = Flask(__name__)

tool_schemas = []
script_dir = os.path.dirname(os.path.abspath(__file__))
for fname in os.listdir(f"{script_dir}/tools"):
    with open(f"{script_dir}/tools/{fname}") as f:
        tool_schemas.append(json.load(f))

LM_STUDIO_URL = "http://localhost:1234/v1"

def start_lm_studio():
    try:
        subprocess.run(["lms", "server", "start"], check=True)
        print("LM Studio server started successfully")
    except subprocess.CalledProcessError:
        print("Failed to start LM Studio server")

def load_model():
    try:
        subprocess.run(["lms", "model", "load", "--first"], check=True)
        print("Model loaded successfully")
    except subprocess.CalledProcessError:
        print("Failed to load model")

def check_lm_studio_status():
    try:
        response = requests.get(f"{LM_STUDIO_URL}/models")
        if response.status_code == 200:
            return True
        return False
    except requests.RequestException:
        return False

def execute_tool(tool_name, tool_args):
    tool_name_to_function = {
        "create_note": create_note,
    }
    if tool_name in tool_name_to_function:
        return tool_name_to_function[tool_name](tool_args)
    else:
        return f"Unknown tool: {tool_name}"

def create_note(tool_args):
    work_notes_path = os.environ.get("ICLOUD_WORK_NOTES_DIR", "")
    intermediate_path = "5 - Unsorted"
    if not work_notes_path:
        return {
            "error": "ICLOUD_WORK_NOTES_DIR environment variable is not set",
            "content": "",
        }
    note_name = tool_args.get("note_name", "Untitled Ori Note")
    note_content = tool_args.get("note_content", "")
    note_path = f"{work_notes_path}/{intermediate_path}/{note_name}.md"

    # Create the file for the note if it doesn't already exist
    if not os.path.exists(note_path):
        with open(note_path, "w") as f:
            f.write(note_content)

    return {
        "error": "",
        "content": f"Created note successfully: {intermediate_path}/{note_name}.md",
    }

@app.route('/<path:subpath>', methods=['POST'])
def proxy(subpath):
    if not check_lm_studio_status():
        start_lm_studio()
        load_model()

    url = f"{LM_STUDIO_URL}/{subpath}"
    response = requests.post(url, json={**request.json, "tools": tool_schemas})

    if response.status_code == 200:
        content = response.json()
        if "tool_calls" in content["choices"][0]["message"]:
            for tool_call in content["choices"][0]["message"]["tool_calls"]:
                tool_name = tool_call["function"]["name"]
                tool_args = json.loads(tool_call["function"]["arguments"])
                tool_result = execute_tool(tool_name, tool_args)
                if tool_result.get("error"):
                    response.status_code = 500
                    content["error"] = tool_result["error"]
                content["tool_results"] = tool_result.get("content", "")
        return jsonify(content)
    else:
        return response.text, response.status_code

if __name__ == '__main__':
    app.run(port=1230)

File: custom-scripts/ori/ori.ts
#!/usr/bin/env bun

import { spawn } from 'child_process';
import { promises as fs } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

interface ToolResult {
  error: string;
  content: string;
}

interface CreateNoteArgs {
  note_name?: string;
  note_content?: string;
}

interface LMStudioResponse {
  choices: Array<{
    message: {
      content?: string;
      role: string;
      tool_calls?: Array<{
        function: {
          name: string;
          arguments: string;
        };
      }>;
    };
    finish_reason?: string;
  }>;
  tool_results?: string;
  error?: string;
}

const LM_STUDIO_URL = 'http://localhost:1234/v1';

// Load tool schemas
const loadToolSchemas = async () => {
  const scriptDir = dirname(fileURLToPath(import.meta.url));
  const toolsDir = join(scriptDir, 'tools');
  const files = await fs.readdir(toolsDir);
  const schemas = await Promise.all(
    files.map(file => fs.readFile(join(toolsDir, file)).then((text) => JSON.parse(text.toString())))
  );
  return schemas;
};

const startLmStudio = (): Promise<void> => {
  return new Promise((resolve, reject) => {
    const process = spawn('lms', ['server', 'start'], { shell: true });
    process.on('close', (code: number) => {
      if (code === 0) {
        console.log('LM Studio server started successfully');
        resolve();
      } else {
        console.error('Failed to start LM Studio server');
        reject(new Error('Failed to start LM Studio server'));
      }
    });
  });
};

const loadModel = (): Promise<void> => {
  return new Promise((resolve, reject) => {
    const process = spawn('lms', ['model', 'load', '--first'], { shell: true });
    process.on('close', (code: number) => {
      if (code === 0) {
        console.log('Model loaded successfully');
        resolve();
      } else {
        console.error('Failed to load model');
        reject(new Error('Failed to load model'));
      }
    });
  });
};

const checkLmStudioStatus = async (): Promise<boolean> => {
  try {
    const response = await fetch(`${LM_STUDIO_URL}/models`);
    return response.status === 200;
  } catch (error) {
    return false;
  }
};

const createNote = async (toolArgs: CreateNoteArgs): Promise<ToolResult> => {
  const workNotesPath = process.env.ICLOUD_WORK_NOTES_DIR;
  const intermediatePath = '5 - Unsorted';
  
  if (!workNotesPath) {
    return {
      error: 'ICLOUD_WORK_NOTES_DIR environment variable is not set',
      content: '',
    };
  }

  const noteName = toolArgs.note_name || 'Untitled Ori Note';
  const noteContent = toolArgs.note_content || '';
  const notePath = join(workNotesPath, intermediatePath, `${noteName}.md`);

  try {
    await fs.writeFile(notePath, noteContent);
    return {
      error: '',
      content: `Created note successfully: ${intermediatePath}/${noteName}.md`,
    };
  } catch (error: any) {
    return {
      error: `Failed to create note: ${error.message}`,
      content: '',
    };
  }
};

const executeTool = async (toolName: string, toolArgs: any): Promise<ToolResult> => {
  const toolNameToFunction: { [key: string]: (args: any) => Promise<ToolResult> } = {
    'create_note': createNote,
  };

  if (toolName in toolNameToFunction) {
    return await toolNameToFunction[toolName](toolArgs);
  }
  return { error: `Unknown tool: ${toolName}`, content: '' };
};

const server = Bun.serve({
  port: 1230,
  async fetch(req) {
    try {
      if (req.method !== 'POST') {
        return new Response('Method not allowed', { status: 405 });
      }

      if (!await checkLmStudioStatus()) {
        await startLmStudio();
        await loadModel();
      }

      const toolSchemas = await loadToolSchemas();
      const body = await req.json() as Record<string, unknown>;
      
      const response = await fetch(`${LM_STUDIO_URL}${new URL(req.url).pathname}`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ ...body, tools: toolSchemas }),
      });

      const content = await response.json() as LMStudioResponse;

      if (response.ok && content.choices?.[0]?.message?.tool_calls) {
        for (const toolCall of content.choices[0].message.tool_calls) {
          content.choices[0].finish_reason = 'stop';
          
          const toolName = toolCall.function.name;
          const toolArgs = JSON.parse(toolCall.function.arguments);
          const toolResult = await executeTool(toolName, toolArgs);

          if (toolResult.error) {
            return Response.json({
              error: toolResult.error,
              choices: [{
                message: {
                  content: toolResult.error,
                  role: 'assistant',
                }
              }]
            }, { status: 500 });
          }

          content.tool_results = toolResult.content;
          content.choices[0].message = {
            content: toolResult.content,
            role: 'assistant',
          };
        }
      }

      return Response.json(content);
    } catch (error: any) {
      return Response.json({ error: error.message }, { status: 500 });
    }
  },
});

console.log(`Server running on port ${server.port}`);
